src/vs/workbench/contrib/notebook/browser/viewModel/notebookOutlineProvider.ts

```diff
@@ -53,29 +54,36 @@ export class NotebookCellOutlineProvider {
 	) {
 		this._outlineEntryFactory = new NotebookOutlineEntryFactory(notebookExecutionStateService);
 
-		const selectionListener = new MutableDisposable();
-		this._disposables.add(selectionListener);
-
-		selectionListener.value = combinedDisposable(
-			Event.debounce<void, void>(
-				_editor.onDidChangeSelection,
-				(last, _current) => last,
-				200
-			)(this._recomputeActive, this),
-			Event.debounce<INotebookViewCellsUpdateEvent, INotebookViewCellsUpdateEvent>(
-				_editor.onDidChangeViewCells,
-				(last, _current) => last ?? _current,
-				200
-			)(this._recomputeState, this)
+		this._disposables.add(Event.debounce<void, void>(
+			_editor.onDidChangeSelection,
+			(last, _current) => last,
+			200
+		)(() => {
+			console.log('1');
+			this._recomputeActive();
+		}, this))
+		this._disposables.add(Event.debounce<INotebookViewCellsUpdateEvent, INotebookViewCellsUpdateEvent>(
+			_editor.onDidChangeViewCells,
+			(last, _current) => last ?? _current,
+			200
+		)(() => {
+			console.log('2');
+			this._recomputeActive();
+		}, this)
 		);
 
+		// .3s of a delay is sufficient, 100-200s is too quick and will unnecessarily block the ui thread.
+		// Given we're only updating the outline when the user types, we can afford to wait a bit.
+		const delayer = this._disposables.add(new Delayer<void>(300));
+		const delayedRecompute = () => delayer.trigger(() => this._recomputeState());
+
 		this._disposables.add(_configurationService.onDidChangeConfiguration(e => {
 			if (e.affectsConfiguration(NotebookSetting.outlineShowMarkdownHeadersOnly) ||
 				e.affectsConfiguration(NotebookSetting.outlineShowCodeCells) ||
 				e.affectsConfiguration(NotebookSetting.outlineShowCodeCellSymbols) ||
 				e.affectsConfiguration(NotebookSetting.breadcrumbsShowCodeCells)
 			) {
-				this._recomputeState();
+				delayedRecompute();
```

Comment:
Debouncing in outline provider resulted in significant improvements
From 60s to run a cell to 2-3s.
(this was already fixed in a previous PR, but this now adds the same debouncing for typing in editor).

When ever we make changes in editor (typing), we trigger outline changes synchronously
As a result, ever key stroke results in outline calculation.
With a large notebook (100 markdown cells and each markdown cell having 10-20 lines), then it takes around 50ms to calculate the outline.
This is ignoring code cells in outline.

Thus every key stroke results in a 50ms delay in renderer.
Solution = debounce